gym_wrapper/base_gym_env.py

```diff
@@ -0,0 +1,241 @@
+            return self.__collect_observations()
+        else:
+            return self.__collect_observations(), {}
+
+    def __collect_observations(self):
+        observation = []
+        if self.state is not None:
+            observation.append(np.transpose(self.state.screen_buffer, (1, 2, 0)))
+            if self.depth:
+                observation.append(self.state.depth_buffer)
+            if self.labels:
+                observation.append(self.state.labels_buffer)
+            if self.position:
+                observation.append(
+                    np.array(
+                        [self.state.game_variables[i] for i in range(4)],
+                        dtype=np.float32,
+                    )
+                )
+                if self.health:
+                    observation.append(
+                        np.array([self.state.game_variables[4]], dtype=np.float32)
+                    )
+            elif self.health:
+                observation.append(
+                    np.array([self.state.game_variables[0]], dtype=np.float32)
+                )
+        else:
+            # there is no state in the terminal step, so a zero observation is returned instead
+            if isinstance(self.observation_space, gym.spaces.box.Box):
+                # Box isn't iterable
+                obs_space = [self.observation_space]
+            else:
+                obs_space = self.observation_space
+
+            for space in obs_space:
+                observation.append(np.zeros(space.shape, dtype=space.dtype))
+
+        # if there is only one observation, return obs as array to sustain compatibility
+        if len(observation) == 1:
+            observation = observation[0]
+        return observation
+
+    def render(self, mode="human"):
+        game_state = self.game.get_state()
+        if game_state is None:
+            img = np.zeros(
+                (
+                    self.game.get_screen_channels(),
+                    self.game.get_screen_height(),
+                    self.game.get_screen_width(),
+                )
+            )
+        else:
+            img = game_state.screen_buffer
+        img = np.transpose(img, [2, 1, 0])
+
+        if self.window_surface is None:
+            pygame.init()
+            pygame.display.set_caption("Vizdoom")
+            if mode == "human":
+                self.window_surface = pygame.display.set_mode(img.shape[:2])
+            else:  # rgb_array
+                self.window_surface = pygame.Surface(img.shape[:2])
+
+        surf = pygame.surfarray.make_surface(img)
+        self.window_surface.blit(surf, (0, 0))
+
+        if mode == "human":
+            pygame.display.update()
+
+        if mode == "rgb_array":
+            return img
+        else:
+            return self.isopen
+
+    def close(self):
+        if self.window_surface:
+            pygame.quit()
+            self.isopen = False
+
+    @staticmethod
+    def get_keys_to_action():
```

Comment:
Hmm is this function used by new Gym APIs or so? I assume it maps keyboard buttons to action indexes.
It might not be true for all scenarios, depending on which order actions were defined in the list :). We might have to skip it for now as I am not sure if there is a reliable way to get the keyboard bindings from vizdoom.

Reply:
No, this is not Gym API. I wasn't sure what to do with it. I have removed it now.